compiler/next/include/chpl/types/QualifiedType.h

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_TYPES_QUALIFIEDTYPE_H
+#define CHPL_TYPES_QUALIFIEDTYPE_H
+
+#include "chpl/types/Type.h"
+
+namespace chpl {
+namespace types {
+
+
+// TODO: is this name too weird given that it includes param values?
+class QualifiedType {
```

Comment:
I think you are asking why I didn't add the fields here as fields on `Type` ? The short answer is that other compilers don't do that (e.g. clang has an idea of a qualified type to handle things like `const int` and it has a different way of representing all of the `int`s). But deeper than that, I think it's useful to be able to know that, say `var x: int` and `const ref y: int` have the same "type" in a way, even though their kinds are different. Also, the QualifiedType here includes a field that has the param values. It would be really hard to work with if we had a different "int" type for every param value.

So what is really going on is that there are two things we are calling a "type":
 * QualifiedType is the "type" used in most cases by resolution and is mostly a useful way to bundle up some compile-time type+param information
 * Type corresponds more to the type the language specification etc talks about
 
I think it makes sense to have separate ways of representing those two things. At the same time, I am open to other names besides QualifiedType.